File: src/adapters/git/GitHubAdapter.ts
import axios from 'axios';

import { GitAdapter } from './GitAdapter';

export interface GitHubPRParams {
  owner: string;
  repo: string;
  pullNumber: string;
}

/**
 * Adapter for GitHub platform
 */
export class GitHubAdapter extends GitAdapter {
  /**
   * Override getLocalDiff to handle CI mode properly for GitHub Actions
   * @param targetBranch - The branch to compare against (ignored in CI mode)
   * @returns The diff content
   */
  override async getLocalDiff(targetBranch?: string): Promise<string> {
    // In CI mode, use GitHub API to fetch PR diff
    if (this.config.mrMode === 'ci') {
      return this.getCIModeFromGitHubActions();
    }
    
    // For non-CI mode, use parent implementation
    return super.getLocalDiff(targetBranch);
  }
  
  /**
   * Get PR diff from GitHub Actions environment
   * @returns The diff content from GitHub API
   */
  private async getCIModeFromGitHubActions(): Promise<string> {
    // GitHub Actions provides these environment variables
    const repository = process.env['GITHUB_REPOSITORY']; // e.g., "owner/repo"
    const eventName = process.env['GITHUB_EVENT_NAME'];
    const ref = process.env['GITHUB_REF']; // e.g., "refs/pull/123/merge"
    
    if (eventName !== 'pull_request') {
      throw new Error(`CI mode is only supported for pull_request events, got: ${eventName}`);
    }
    
    if (!repository) {
      throw new Error('GITHUB_REPOSITORY environment variable not found');
    }
    
    if (!ref || !ref.startsWith('refs/pull/')) {
      throw new Error(`Invalid GITHUB_REF for pull request: ${ref}`);
    }
    
    // Extract PR number from ref: refs/pull/123/merge -> 123
    const prMatch = ref.match(/refs\/pull\/(\d+)\/merge/);
    if (!prMatch) {
      throw new Error(`Could not extract PR number from GITHUB_REF: ${ref}`);
    }
    
    const repositoryParts = repository.split('/');
    if (repositoryParts.length !== 2) {
      throw new Error(`Invalid GITHUB_REPOSITORY format: ${repository}`);
    }
    
    const [owner, repo] = repositoryParts;
    const pullNumber = prMatch[1];
    
    if (!owner || !repo || !pullNumber) {
      throw new Error(`Failed to parse GitHub repository information: owner=${owner}, repo=${repo}, pullNumber=${pullNumber}`);
    }
    
    // Use existing getRequestDiff method
    return this.getRequestDiff({ owner, repo, pullNumber });
  }
  /**
   * Parse a GitHub PR URL to extract owner, repo, and PR number
   * @param url - The URL to parse
   * @returns Object containing owner, repo, and pullNumber
   * @throws If URL is invalid or cannot be parsed
   */
  parseRequestUrl(url: string): Record<string, string> {
    try {
      const parsedUrl = new URL(url);

      // Expected format: https://github.com/owner/repo/pull/123
      const pathParts = parsedUrl.pathname.split('/');

      if (pathParts.length < 5 || pathParts[3] !== 'pull') {
        throw new Error('Invalid GitHub PR URL format');
      }

      const owner = pathParts[1] ?? '';
      const repo = pathParts[2] ?? '';
      const pullNumber = pathParts[4] ?? '';

      return { owner, repo, pullNumber };
    } catch (error) {
      console.error('Error parsing GitHub PR URL:', (error as Error).message);
      throw new Error(`Invalid GitHub PR URL: ${url}`);
    }
  }

  /**
   * Get the diff of a pull request from GitHub
   * @param params - Parameters with owner, repo, and pullNumber
   * @returns The diff content
   */
  async getRequestDiff(params: Record<string, string>): Promise<string> {
    const ghParams = params as unknown as GitHubPRParams;
    try {
      // GitHub API returns the diff directly with the right Accept header
      const url = `https://api.github.com/repos/${ghParams.owner}/${ghParams.repo}/pulls/${ghParams.pullNumber}`;

      const response = await axios.get(url, {
        headers: {
          Authorization: `Bearer ${(this as any).config.githubToken || ''}`,
          Accept: 'application/vnd.github.v3.diff',
          'X-GitHub-Api-Version': '2022-11-28',
        },
      });

      return response.data;
    } catch (error) {
      console.error('Error fetching PR diff from GitHub:', (error as Error).message);
      if (axios.isAxiosError(error) && error.response) {
        console.error('Response status:', error.response.status);
        console.error('Response data:', error.response.data);
      }
      throw error;
    }
  }

  /**
   * Post a comment on a GitHub pull request
   * @param params - Parameters with owner, repo, and pullNumber
   * @param comment - The comment content
   */
  async commentOnRequest(params: Record<string, string>, comment: string): Promise<void> {
    const ghParams = params as unknown as GitHubPRParams;
    try {
      const url = `https://api.github.com/repos/${ghParams.owner}/${ghParams.repo}/issues/${ghParams.pullNumber}/comments`;

      await axios.post(
        url,
        { body: `## LLM Code Review Feedback\n\n${comment}` },
        {
          headers: {
            Authorization: `Bearer ${(this as any).config.githubToken || ''}`,
            Accept: 'application/vnd.github.v3+json',
            'X-GitHub-Api-Version': '2022-11-28',
          },
        }
      );
    } catch (error) {
      console.error('Error posting comment on PR:', (error as Error).message);
      if (axios.isAxiosError(error) && error.response) {
        console.error('Response status:', error.response.status);
        console.error('Response data:', error.response.data);
      }
      throw error;
    }
  }
}

File: python/tiw/adapters/git/github_adapter.py
"""GitHub API adapter for PR operations."""

import re
from typing import Dict, Optional

from github import Github
from github.GithubException import GithubException

from ...config.config import AppConfig
from .git_adapter import GitAdapter


class GitHubAdapter(GitAdapter):
    """Adapter for GitHub API."""
    
    def __init__(self, config: AppConfig):
        """Initialize the GitHub adapter.
        
        Args:
            config: Application configuration
        """
        super().__init__(config)
        
        # Initialize the client
        self.github = self._init_client()
    
    def _init_client(self) -> Github:
        """Initialize the GitHub API client.
        
        Returns:
            GitHub client
            
        Raises:
            ValueError: If token is missing
        """
        if not self.config.github_token:
            raise ValueError("GitHub token is required")
        
        return Github(self.config.github_token)
    
    def parse_request_url(self, url: str) -> Dict[str, str]:
        """Parse a GitHub PR URL into owner, repo, and PR number.
        
        Args:
            url: The GitHub PR URL
            
        Returns:
            Dictionary with owner, repo, and PR number
            
        Raises:
            ValueError: If URL is invalid
        """
        # GitHub PR URL format: https://github.com/owner/repo/pull/123
        pattern = r'https?://github\.com/([^/]+)/([^/]+)/pull/(\d+)'
        match = re.match(pattern, url)
        
        if not match:
            raise ValueError(f"Invalid GitHub PR URL: {url}")
        
        return {
            "owner": match.group(1),
            "repo": match.group(2),
            "pr_number": match.group(3),
        }
    
    async def get_request_diff(self, params: Dict[str, str]) -> str:
        """Get the diff of a GitHub PR.
        
        Args:
            params: Parameters needed to identify the PR (owner, repo, pr_number)
            
        Returns:
            The diff content
            
        Raises:
            ValueError: If required parameters are missing
            Exception: If GitHub API request fails
        """
        # Validate required params
        if not all(key in params for key in ["owner", "repo", "pr_number"]):
            raise ValueError("Missing required parameters: owner, repo, pr_number")
        
        try:
            # Get repository and pull request
            repo = self.github.get_repo(f"{params['owner']}/{params['repo']}")
            pr = repo.get_pull(int(params["pr_number"]))
            
            # Get the diff
            # GitHub API doesn't provide diff directly, so we'll fetch all files changed
            files = pr.get_files()
            
            # Format as a unified diff
            diff_content = []
            for file in files:
                if self.config.ignore_lock_files and self._is_lock_file(file.filename):
                    continue
                
                if file.status == "removed":
                    # File was deleted
                    diff_content.append(f"diff --git a/{file.filename} b/{file.filename}")
                    diff_content.append(f"deleted file mode 100644")
                    diff_content.append(f"--- a/{file.filename}")
                    diff_content.append(f"+++ /dev/null")
                    if file.patch:
                        diff_content.append(file.patch)
                elif file.status == "added":
                    # File was added
                    diff_content.append(f"diff --git a/{file.filename} b/{file.filename}")
                    diff_content.append(f"new file mode 100644")
                    diff_content.append(f"--- /dev/null")
                    diff_content.append(f"+++ b/{file.filename}")
                    if file.patch:
                        diff_content.append(file.patch)
                else:
                    # File was modified
                    diff_content.append(f"diff --git a/{file.filename} b/{file.filename}")
                    diff_content.append(f"--- a/{file.filename}")
                    diff_content.append(f"+++ b/{file.filename}")
                    if file.patch:
                        diff_content.append(file.patch)
            
            return "\n".join(diff_content)
        
        except GithubException as e:
            raise Exception(f"GitHub API error: {e.status} - {e.data.get('message', str(e))}")
        except Exception as e:
            raise Exception(f"Error getting PR diff: {str(e)}")
    
    def _is_lock_file(self, filename: str) -> bool:
        """Check if a file is a lock file.
        
        Args:
            filename: Filename to check
            
        Returns:
            True if it's a lock file
        """
        return filename.endswith(('.lock', 'lock.json', 'package-lock.json', 'yarn.lock', 'pnpm-lock.yaml'))
    
    async def comment_on_request(self, params: Dict[str, str], comment: str) -> None:
        """Post a comment on a GitHub PR.
        
        Args:
            params: Parameters needed to identify the PR (owner, repo, pr_number)
            comment: The comment content
            
        Raises:
            ValueError: If required parameters are missing
            Exception: If GitHub API request fails
        """
        # Validate required params
        if not all(key in params for key in ["owner", "repo", "pr_number"]):
            raise ValueError("Missing required parameters: owner, repo, pr_number")
        
        try:
            # Get repository and pull request
            repo = self.github.get_repo(f"{params['owner']}/{params['repo']}")
            pr = repo.get_pull(int(params["pr_number"]))
            
            # Post the comment
            pr.create_issue_comment(comment)
        
        except GithubException as e:
            raise Exception(f"GitHub API error: {e.status} - {e.data.get('message', str(e))}")
        except Exception as e:
            raise Exception(f"Error posting comment: {str(e)}")